File: 数据结构/mergeSort.py
# 1.
def mergeSort(alist):
    if len(alist)<=1:
        return alist
    mid = len(alist)//2
    left = mergeSort(alist[:mid])
    right = mergeSort(alist[mid:])
    return merge(left, right)

def merge(left, right):
    newlist = []
    i=0
    j=0
    while i<len(left) and j<len(right):
        if left[i]<right[j]:
            newlist.append(left[i])
            i+=1
        else:
            newlist.append(right[j])
            j+=1
    newlist+=left[i:]
    newlist+=right[j:]
    return newlist


# 2. directly operate on `alist`
def mergeSort(alist):
    if len(alist) > 1:
        mid = len(alist)//2
        lefthalf = alist[:mid]
        righthalf = alist[mid:]

        mergeSort(lefthalf)
        mergeSort(righthalf)

        i = 0; j = 0; k = 0
        while i < len(lefthalf) and j < len(righthalf):
            if lefthalf[i] < righthalf[j]:
                alist[k] = lefthalf[i]
                i += 1
            else:
                alist[k] = righthalf[j]
                j += 1
            k += 1

        while i < len(lefthalf):
            alist[k] = lefthalf[i]
            i += 1
            k += 1
        while j < len(righthalf):
            alist[k] = righthalf[j]
            j += 1
            k += 1

alist = [3,2,5,8,6,9,4,7,]
print mergeSort(alist)

File: 数据结构/merge_sort.cpp
//归并是先分后排，归并的分界点在定义域（下标）上
//快排是先排后分,快排的分界点在值域上
#include <iostream>
using namespace std;

const int N = 1e5;
int n, q[N], tmp[N];

void mergeSort(int q[], int l, int r){
    if (l==r)   return;
    
    int mid = l + r >> 1;
    //先分
    mergeSort(q, l, mid);   mergeSort(q, mid+1, r);
    //再合
    int k = 0, i = l, j = mid + 1;
    while(i <= mid && j <= r){
        if (q[i]<=q[j]) tmp[k++] = q[i++];
        else            tmp[k++] = q[j++];
    }
    //第一组还有数
    while(i <= mid)     tmp[k++] = q[i++];
    //第二组还有数
    while(j <= r)       tmp[k++] = q[j++];
    //tmp换回q
    //i指向q[] j指向tmp[]
    for(i=l, j=0; i<=r; ++i, ++j)   q[i] = tmp[j];
}


int main(){
    scanf("%d", &n);
    for (int i = 0; i < n; ++ i)    scanf("%d", &q[i]);
    mergeSort(q, 0, n-1);
    for (int i = 0; i < n; ++ i)    printf("%d ", q[i]);
    
    return 0;
}
